rangeproof/proof.go

```diff
@@ -0,0 +1,463 @@
+//
+// Since k > m, and since d1, d2, d3 are real, we have d1^2+d2^2+d3^2+k-m > 0. Now either phi(n) divides d1^2+d2^2+d3^2+k-m, or
+// v5-d1*v1+d2*v2+d3*v3 is not zero. If phi(n) divides d1^2+d2^2+d3^2+k-m, then by Lemma 11 in CL03, we can factor n and trivially
+// solve the instance.
+//
+// Otherwise, we now have nonzero a = d1^2+d2^2+d3^2+k-m, and b = v5-d1*v1+d2*v2+d3*v3, such that R^a = S^b. Since e does not divide phi(n)
+// (otherwise, we could just factor n again, this time by lemma 12 in CL03), this also implies u^(r1*a) = u^b. Since a and b are bounded
+// (each of their components is proven to be smaller than a bound during the ZKP), we can take r1 large enough to guarantee r1*a > b.
+// Then r1*a-b > 0, and by multiplication by u^-b, we get u^(r1*a-b) = 1, hence phi(n) | r1*a-b, which means we can factor n and use
+// that to solve the flexible RSA problem.
+//
+// This shows existence of G with non-neglible succes probability also contradicts strong RSA, hence existence of A contradicts
+// the strong RSA assumption.
+//
+// The techniques used to fake issuance can be generalized to multiple issuances using techiques similar to those used in CL03 for lemmas 3-5,
+// the rest of the proof then needs e to be replaced with E, the product of all used e_i's.
+
+type (
+	ProofStructure struct {
+		cRep     []qrRepresentationProofStructure
+		mCorrect qrRepresentationProofStructure
+
+		a int
+		k *big.Int
+
+		splitter Splitter
+		ld       uint
+		lm       uint
+		lh       uint
+		lstatzk  uint
+	}
+
+	Proof struct {
+		// Actual proof responses
+		C                 []*big.Int `json:"C"`
+		DResponse         []*big.Int `json:"d"`
+		VResponse         []*big.Int `json:"v"`
+		VCombinedResponse *big.Int   `json:"v5"`
+		MResponse         *big.Int   `json:"m"`
+	}
+
+	ProofCommit struct {
+		// Bases
+		c []*big.Int
+
+		// Secrets
+		d            []*big.Int
+		dRandomizer  []*big.Int
+		v            []*big.Int
+		vRandomizer  []*big.Int
+		v5           *big.Int
+		v5Randomizer *big.Int
+		m            *big.Int
+		mRandomizer  *big.Int
+	}
+
+	proof       Proof
+	proofCommit ProofCommit
+	qrGroup     QrGroup
+)
+
+// Create a new proof structure for proving a statement of form a*m - k >= 0
+//  splitter describes the method used for splitting numbers into sum of squares
+//  lh is the size of the challenge
+//  lm the size of m, and also used as the number of bits for computational hiding
+//  lstatzk the number of bits of statistical hiding to use
+func New(a int, k *big.Int, split Splitter, lh, lstatzk, lm uint) *ProofStructure {
+	if split.Nsplit() > 4 {
+		panic("No support for range proofs with delta split in more than 4 squares")
+	}
+
+	if split.Nsplit() == 3 {
+		// Not all numbers can be written as sum of 3 squares, but n for which n == 2 (mod 4) can
+		// so ensure that a*m-k falls into that category
+		a *= 4
+		k = new(big.Int).Mul(k, big.NewInt(4)) // ensure we dont overwrite callers copy of k
+		k.Sub(k, big.NewInt(2))
```

Comment:
This works because the proof proves a*m-k >= 0, not a*m-k > 0. This guarantees 4*a*m - (4*k-2) >= 0 iff a*m-k >= 0

Reply:
so there is no modification of the statement because of this parameter substitution

Reply:
Alright, great.

In that case, however, this package doesn't do what the comment on top of this file claims it does, as in that comment `>` is used multiple times instead of `>=`, so then the comment should be fixed.